### ISSUE TITLE
Making modifications to an individual in a genetic algorithm

### ISSUE THREAD

**Issue by pranavdheram** [DEAP/deap#182]
Hello! I am trying to employ genetic algorithms to optimize Markowitz's 'Portfolio Optimization model'. Each individual is a set of weights corresponding to the weight of a stock in a portfolio. Therefore these weights must sum to one. However, after crossover and mutation, I have no way of ensuring that this condition is met. 
This is what I have tried doing: 
Let a population be a list of lists(i.e, a list of individuals)
So, after crossover and mutation, I do this : 
offspring = [[float(j)/sum(i) for j in i] for i in offspring]
where, offspring is the list of individuals. 
so basically, this : individual = individual/sum(individual)

However, when I evaluate these individuals, I encounter an error: 'list' object has no attribute 'fitness'. 
I know why this is happening. When we call the creator it associates a variable called fitness with each individual(list). So, in the above operation, this initialized individual is lost. 

Is there a way around this? Basically, some way through which I can ensure that the weights of my individual sum to 1 after crossover and mutation also? 

Thank you very much for your time.

**Comment by Ginfung** [comment]
one idea is copy the source code of GA, and re-normalize your candidates after crossover and mutation.

**Comment by pranavdheram** [comment]
Thank you for your prompt reply. The problem with renormalizing the individuals in the population after crossover and mutation is, as I had mentioned earlier, I end up creating a new list in the process. However, from what I understand, the GA creates an individual as an object of sorts, with an attribute called 'fitness.valid'. 
So when i do this after crossover and mutation in an attempt to re-normalize:
offspring = [[float(j)/sum(i) for j in i] for i in offspring]
I essentially create a new list called 'offspring' which no longer has the 'fitness.valid' attribute which is essential during evaluation. 
Isn't there a way to access each individual in the population and further divide each element of the individual by the sum of the individual? This way, I won't be creating a new list. 
Thanks.

**Comment by fmder** [comment]
You can replace the content of an individual without changing its type with

for ind in pop:
    ind[:] = ind/sum_



Le 23 mars 2017 05:26, "pranavdheram" <notifications@github.com> a écrit :

Thank you for your prompt reply. The problem with renormalizing the
individuals in the population after crossover and mutation is, as I had
mentioned earlier, I end up creating a new list in the process. However,
from what I understand, the GA creates an individual as an object of sorts,
with an attribute called 'fitness.valid'.
So when i do this after crossover and mutation in an attempt to
re-normalize:

offspring = [[float(j)/sum(i) for j in i] for i in offspring]
I essentially create a new list called 'offspring' which no longer has the
'fitness.valid' attribute which is essential during evaluation.
Isn't there a way to access each individual in the population and further
divide each element of the individual by the sum of the individual? This
way, I won't be creating a new list.
Thanks.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/DEAP/deap/issues/182#issuecomment-288662096>, or mute
the thread

**Comment by pranavdheram** [comment]
That worked!! Thank you so much.
